File: client/src/main/kotlin/dev/throwouterror/game/client/engine/Ray.java
package dev.throwouterror.game.client.engine;

import dev.throwouterror.util.math.Tensor;

public class Ray {

    private Tensor direction = new Tensor();
    private Tensor origin = new Tensor();

    public Tensor getDirection() {
        return direction;
    }

    public Ray setDirection(float x, float y, float z) {
        this.direction.setValues(x, y, z);
        return this;
    }

    public Ray setDirection(Tensor direction) {
        this.direction = direction.clone();
        return this;
    }

    public Tensor getOrigin() {
        return origin;
    }

    public Ray setOrigin(float x, float y, float z) {
        this.origin.setValues(x, y, z);
        return this;
    }

    public Ray setOrigin(Tensor origin) {
        this.origin = origin.clone();
        return this;
    }

    @Override
    public String toString() {
        return "(origin: " + origin + " direction: " + direction + ")";
    }

}

File: client/src/main/kotlin/dev/throwouterror/game/client/engine/Ray.kt
package dev.throwouterror.game.client.engine

import dev.throwouterror.util.math.Tensor

class Ray {
    var direction = Tensor.VECTOR_ZERO.clone()
        private set
    var origin = Tensor.VECTOR_ZERO.clone()
        private set

    fun setDirection(x: Float, y: Float, z: Float): Ray {
        this.direction.set(x, y, z)
        return this
    }

    fun setDirection(direction: Tensor): Ray {
        this.direction = direction.clone()
        return this
    }

    fun setOrigin(x: Float, y: Float, z: Float): Ray {
        this.origin.set(x, y, z)
        return this
    }

    fun setOrigin(origin: Tensor): Ray {
        this.origin = origin.clone()
        return this
    }

    public override fun toString(): String {
        return "(origin: $origin direction: $direction)"
    }
}
